### PR TITLE
Add code formatters & linters

### PR DESCRIPTION
**Pull request by jack60612**
Added black (code formatter), isort (import formatter) and flake8 a pep8 abiding code linter 
Also added gh actions workflows for everything above

### CODE REVIEW

**Comment by jack60612** [comment]
@virtuald where would you want me to put the information on how to use the formatters & linters?

**Comment by auscompgeek** [comment]
We already run black in CI?

**Comment by jack60612** [comment]
> We already run black in CI?

oops, im a little blind

**Comment by jack60612** [comment]
> > We already run black in CI?
> 
> oops, I must have missed it.

Wait, why did it reformat things then?

**Comment by jack60612** [comment]
What should i use as the standardize

> > > We already run black in CI?
> > 
> > 
> > oops, I must have missed it.
> 
> ~Wait, why did it reformat things then?~ Line length.

I set the line length to 120 because it seemed like that was what was being used, but i can change that back to 88 if thats whats wanted.

File: docs/conf.py [hunk 1]
```diff
@@ -5,10 +5,12 @@
 # Imports
 #
 
-import sys
 import os
 from os.path import abspath, dirname
 
+from robotpy_sphinx.regen import gen_package
```
Comment (virtuald) [inline docs/conf.py:11]: No, this is a configuration file, it's easier to have things grouped together.

File: pyproject.toml [hunk 2]
```diff
@@ -360,3 +360,10 @@ SingleJointedArmSim = "frc/simulation/SingleJointedArmSim.h"
 SolenoidSim = "frc/simulation/SolenoidSim.h"
 UltrasonicSim = "frc/simulation/UltrasonicSim.h"
 XboxControllerSim = "frc/simulation/XboxControllerSim.h"
+
+[tool.black]
+line-length = 120
```
Comment (virtuald) [inline pyproject.toml:365]: No configuration is needed for black, the defaults are fine.

File: wpilib/__init__.py [hunk 3]
```diff
@@ -1,17 +1,24 @@
...
+
 # autogenerated by 'robotpy-build create-imports wpilib'
 from ._wpilib import (
     ADIS16448_IMU,
     ADIS16470_IMU,
     ADXL345_I2C,
     ADXL345_SPI,
     ADXL362,
-    ADXRS450_Gyro,
+    CAN,
```
Comment (virtuald) [inline wpilib/__init__.py:14]: Why is this sort not alphabetical? That's just weird.

**Review by virtuald — CHANGES_REQUESTED** [review]
This feels like formatting changes just for the sake of formatting changes. Let's not do that, and it seems isort's sort is terrible.

Did flake8 find anything useful? It's hard to tell with all the formatting changes.

File: wpilib/__init__.py (hunk 3 above)
Comment (jack60612) [reply wpilib/__init__.py:14]: huh isort is supposed to be alphabetical, that is weird.

Comment (zacbir) [reply wpilib/__init__.py:14]: My guess is `ALL_CAPS` > `MixedCaps` > `all_lower`

**Comment by kaechele** [comment]
Would it make sense to use [pre-commit](https://pre-commit.com/) to front-load any formatting changes? Only being able to commit formatted code improves legibility of diffs commit to commit.

**Comment by jack60612** [comment]
oops, i got distracted by the season, ill look at this again.

**Comment by virtuald** [comment]
> Would it make sense to use [pre-commit](https://pre-commit.com/) to front-load any formatting changes? Only being able to commit formatted code improves legibility of diffs commit to commit.

Just enable format on save in your editor?

**Comment by kaechele** [comment]
> Just enable format on save in your editor?

I think that's pretty standard nowadays.

What I was suggesting is different from that though.
Using pre-commit can help enforce repository-wide linting and style rules for all future contributions and allows contributors to catch any linting or formatting errors upon committing code locally and before they are pushed and subsequently fail in CI.
Since the pre-commit hooks are installed by each contributor after checking out the repository people can choose whether they'd like that convenience or not.
Configuring the linting and style rules for this purpose also creates explicit guidelines of how the project expects code to be formatted.

**Comment by virtuald** [comment]
If you want to go down this path again, we've moved everything to the mostwpilib repository.

**Pull request closed without merging**